Add support for Uefi to various fileio libraries

Zitt opened the pull request.

Zitt committed:
micropython-lib/glob: Add uefi support

Add uefi support for internal re engine on uefi port.

Signed-off-by: John Zitterkopf <john.zitterkopf@intel.com>

Zitt committed:
micropython-lib/fnmatch: add uefi support

Add uefi support for fnmatch so it uses internal uefi _re engine.

Signed-off-by: John Zitterkopf <john.zitterkopf@intel.com>

Zitt committed:
micropython-lib/fnmatch: add uefi support

Add uefi support for fnmatch so it uses internal uefi _re engine.
Peer Review changes from PR#418

Signed-off-by: John Zitterkopf <john.zitterkopf@intel.com>

Zitt committed:
micropython-lib/os.path: Add uefi support

Add support for uefi paths in the os.path modules. Borrows code from
ntpath implementation in cPython. Uses internal uefi _re engine.

Signed-off-by: John Zitterkopf <john.zitterkopf@intel.com>

Zitt committed:
micropython-lib/os.path: Add uefi support

Add support for uefi paths in the os.path modules. Borrows code from
ntpath implementation in cPython. Uses internal uefi _re engine.

Signed-off-by: John Zitterkopf <john.zitterkopf@intel.com>

stinos commented on glob/glob.py:
```diff
@@ -2,7 +2,15 @@
-import re
+import sys
+
+if (sys.platform == 'uefi') and (sys.implementation.name == 'micropython'):
```
Canonical way of doing this is
```
try
  import _re
except ImportError:
  import ure
except importError:
  import re
```
which doesn't have the downside of hardcoding these names.
However, the `import re` was there already: why is it now needed to try the `ure` import (not related to uefi I assume)?

Zitt replied:
Good point regarding re vs ure. I guess my thought was *if* someone were to try and run the micropython under cpython; wouldn't it be better to fall back to standard re which would be available on that platform?

stinos replied:
If you look in the source code, using the try/except is the canonical way to do this for MicroPython so it is only vague if you haven't seen any of the other code. Though I get what you mean. Still, I don't think we can just go changing these `import re` everywhere. Does it not work for you?

stinos commented on fnmatch/fnmatch.py:
```diff
@@ -26,10 +35,10 @@ def fnmatch(name, pat):
     [!seq]  matches any char not in seq
 
-    An initial period in FILENAME is not special.
+    An initial period in FILENAME is not special
```
Unrelated formatting would better go in separate sommits. Thoug I'm not sure this is even need?

Zitt replied:
I didn't specifically add the 'period' at the end; but it could have been a typo on my part.

stinos commented on fnmatch/fnmatch.py:
```diff
@@ -50,9 +59,15 @@ def filter(names, pat):
-        if match(os.path.normcase(name)):
-            result.append(name)
+    if os.path is posixpath:
+        # normcase on posix is NOP. Optimize it away from the loop.
```
Have you measured if this has an actual impact?

Zitt replied:
I didn't specifically make this change either. Wonder what is going on.

stinos commented on fnmatch/fnmatch.py:
```diff
@@ -64,6 +79,29 @@ def fnmatchcase(name, pat):
+def _re_translate(p, chmap):
+    oldp = p
+    for idx,c in enumerate(p):
+     print('[%d]=%s' % (idx, c))
```
Leftover? Also please remove the other debug statements.

Zitt replied:
Correct.

stinos commented on fnmatch/fnmatch.py:
```diff
@@ -64,6 +79,29 @@ def fnmatchcase(name, pat):
+    #print( '%s -> %s' % (oldp, p) )
+    return p
+
+def _re_escape(pattern):
```
Would be best to add tests for this.

Zitt replied:
I'm not familiar with tests; would need someone to assist on writing tests.

stinos replied:
See test_fnmatch.py in this directory

stinos commented on fnmatch/fnmatch.py:
```diff
@@ -64,6 +79,29 @@ def fnmatchcase(name, pat):
+    """
+    print( _re_special_chars_map )
+    if isinstance(pattern, str):
+        #return pattern.translate(_re_special_chars_map)
```
dead code

Zitt replied:
Why do you believe this is dead code? what if pattern is a bytestring?
Again; this code is present in cpython IIRC.

stinos replied:
I'm talking about the comment: doesn't do anything, unclear why it is commented out.

stinos commented on fnmatch/fnmatch.py:
```diff
@@ -64,6 +79,29 @@ def fnmatchcase(name, pat):
+    """
+    Escape special characters in a string.
+    """
+    print( _re_special_chars_map )
```
not needed

stinos commented on fnmatch/fnmatch.py:
```diff
@@ -64,6 +79,29 @@ def fnmatchcase(name, pat):
+     print('[%d]=%s' % (idx, c))
+     if ord(c) in chmap:
+      newp = p.replace(c, chmap[ord(c)])
+      #print( '%s -> %s' % (oldp, newp) )
```
remve this and return in one line

stinos commented on fnmatch/fnmatch.py:
```diff
@@ -64,6 +79,29 @@ def fnmatchcase(name, pat):
+        #return pattern.translate(_re_special_chars_map)
+        return _re_translate(pattern, _re_special_chars_map )
+    else:
+        pattern = str(pattern, 'latin1')
```
Hardcoding an encoding might not be ok here? And why latin1?

Zitt replied:
A copy from cpython implementation. Felt is was better to leave as-is; rather than risk a bug later if/when encoding is supported.

stinos replied:
Is it ok to just copy from CPython without acnkowledging that? What does its license say?

stinos commented on fnmatch/fnmatch.py:
```diff
@@ -99,6 +137,11 @@ def translate(pat):
                     stuff = '\\' + stuff
                 res = '%s[%s]' % (res, stuff)
         else:
+         if (sys.platform == 'uefi') and (sys.implementation.name == 'micropython') :
```
Probably niver t make this a 'constant' at the top of the file. However, is this check always correct?

Zitt replied:
This code is correct; all though; there seems to be some indentation issues. 
I'll re-check on my master file. That said the "else" goes with the 
`elif stuff[0] == '^':`
above it so that the _re_escape(c) is called in the case of uefi implementation.

stinos commented on os.path/os/path.py:
```diff
@@ -1,7 +1,38 @@
+ sep = '\\'
+ altsep = '/'
+ import _re as re
+else: import ure as re
```
Code style not consistent.

Zitt replied:
I don't follow. What code style am I violating?

stinos replied:
Well, MicroPython uses black for formatting. But only for the micropython repository currently, not yet for micropython-lib. In any case: all statements should go on their own line for consistency and readability

stinos commented on os.path/os/path.py:
```diff
@@ -1,7 +1,38 @@
+    else:
+        return '\\/'
+
+def isabs(s):
```
Tests would be nice.

stinos commented on os.path/os/path.py:
```diff
@@ -1,7 +1,38 @@
 sep = "/"
+altsep = "\\"
+if (sys.platform == 'uefi') and (sys.implementation.name == 'micropython'):
+ sep = '\\'
```
This is actually also true for windows. Does this mean all the uefi stuff here is appliccable for windows as well?

Zitt replied:
Possibly; however, I don't have a specific need for windows at the moment? Is there a developer who wants to take on windows?

stinos replied:
I almost exclusively use Windows so if there's a major effort towards getting this path problem fixed it would be nice and I could contribute. That's also why I think the 'if platform is X and something else is Y' should be abstracted away in one place instead of repeating. However: in the end I just resolved to changing all backslahes into forward slashes whenevr I use paths: IIRC there are other places apart from os.path and fnmatch which have problems with backslashes.

Zitt committed:
Merge branch 'master' of https://github.com/Zitt/micropython-lib

Zitt committed:
micropython-lib/configparser: Add configparser

Add support for configparser from micropython-lib/pull/265
https://github.com/micropython/micropython-lib/pull/265

Signed-off-by: John Zitterkopf <john.zitterkopf@intel.com>

Zitt committed:
micropython-lib/os, micropython-lib/os.path: better uefi Support

Add more support for uefi paths in the os.path module.
Move os.path into os module to make it happy.

Signed-off-by: John Zitterkopf <john.zitterkopf@intel.com>

dpgeorge commented:
Are you still interested in this PR?  It will need some work to rebase it on the latest master commit.
